Move DependencyModel to libraries

eerhardt opened the pull request:
Moving DependencyModel to the same folder and infrastructure as the rest of our libraries, and out of the installer folder.

I also dropped support for building live anything below netstandard2.0 so we don't need to reference Newtonsoft. The lower TFMs are harvested from the previously shipped package.

Contributes to #3470
Fix #3425

eerhardt committed:
Move DependencyModel to libraries

Moving DependencyModel to the same folder and infrastructure as the rest of our libraries, and out of the installer folder.

I also dropped support for anything below netstandard2.0 at this time.

Contributes to #3470
Fix #3425

eerhardt commented:
error:

```
C:\Users\eerhardt\.nuget\packages\microsoft.dotnet.genfacades\5.0.0-beta.20171.1\build\Microsoft.DotNet.GenPartialFacadeSource.targets(30,5):
error : The type 'System.Collections.Generic.CollectionExtensions' is defined in multiple seed assemblies.
The multiple assemblies are Microsoft.Extensions.DependencyModel.dll, System.Collections.dll. If this is intentional, specify the alias for this type and project reference 
[F:\git\runtime\src\libraries\shims\generated\netstandard.csproj]
```

This is unfortunate that we shipped this API already. Would this be OK to make a breaking change in `5.0` for this assembly? We are already making other breaking changes -
* dropping support for `net451`, `netstandard1.x`
* In `3.0`, we accidentally shipped an unintentional public type in this assembly: 

https://github.com/dotnet/runtime/blob/335ffb05fbe97187452046bc6791405ea33f210c/src/installer/managed/Microsoft.Extensions.DependencyModel/Microsoft.Extensions.DependencyModel.csproj#L11

However, this would be a binary and potentially source breaking change to change `CollectionExtensions`, but I think it might be worth it going forward.

Thoughts - @ericstj @joperezr @safern ?

eerhardt committed:
Add missing copyright headers.

ericstj commented:
We found a similar bug in other extensions libraries #33998.  Since its just extension methods that folks aren't going to directly reference it's not a terrible type conflict.  I'd just suppress the error for now and file a bug.  If folks have a scenario for using the two types and notice the clash we can address it.
https://github.com/dotnet/runtime/blob/562232c2f7ac24551abdfab04c1186740fcb4e50/src/libraries/pkg/test/packageTest.targets#L24-L25

> HashCodeCombiner

What do you suggest?  Removing is breaking, we could obsolete it and/or implement on top of System.HashCode?

joperezr commented on src/libraries/Microsoft.Extensions.DependencyModel/Directory.Build.props:
```diff
@@ -0,0 +1,7 @@
+﻿<Project>
+  <Import Project="..\Directory.Build.props" />
+  <PropertyGroup>
+    <StrongNameKeyId>MicrosoftAspNetCore</StrongNameKeyId>
```
I'm not seeing where was this set on the installer but I trust that maybe some common files where setting it?

eerhardt replied:
It is set here:

https://github.com/dotnet/runtime/blob/15abeaa93110f0d06b741658ba47e75d7584dd9e/src/installer/managed/CommonManaged.props#L31-L32

safern replied:
I think @maryamariyan is adding logic in `Directory.Build.props` when the project name starts with `Microsoft.Extensions` to set the `StrongNameKeyId` to `MicrosoftAspNetCore`

eerhardt replied:
@safern - That doesn't appear to be working. I think it is getting overwritten somewhere. On my machine all the other `Microsoft.Extensions.*` assemblies have `PublicKeyToken=31bf3856ad364e35`. `MicrosoftAspNetCore`'s public key token is `adb9793829ddae60`.

ericstj replied:
We recently fixed this.  There was an ordering bug in the props.  Make sure you pull latest master.

ericstj replied:
https://github.com/dotnet/runtime/commit/87e7991b847a1c80654e05d81005f0931a04ab71

maryamariyan replied:
I believe all should work fine without this file.

eerhardt replied:
Ok, I have removed this file now that I sync'd with master.

joperezr commented on src/libraries/Microsoft.Extensions.DependencyModel/Directory.Build.props:
```diff
@@ -0,0 +1,7 @@
+﻿<Project>
+  <Import Project="..\Directory.Build.props" />
+  <PropertyGroup>
```
Do we just want to follow the Assembly version that the rest of libraries are using(5.0.0.0)? Or do we want to override the default to match what the installer had before?

eerhardt replied:
The assembly version of the "core-setup" managed libraries has followed the versioning of the .NET Core product as a whole - the same as the `dotnet.exe` host.

The 3.1.3 package has:

![image](https://user-images.githubusercontent.com/8291187/78177451-c3731d00-7423-11ea-9cfa-a19488af8a05.png)

(Why we bumped the patch AssemblyVersion, I don't know.)

The 5.0.0-preview.1.20120.5 package has:

![image](https://user-images.githubusercontent.com/8291187/78177538-e69dcc80-7423-11ea-8f60-4ca1be513428.png)

So continuing with `5.0.0.0` will work for this library.

joperezr commented on src/libraries/Microsoft.Extensions.DependencyModel/src/Microsoft.Extensions.DependencyModel.csproj:
```diff
@@ -0,0 +1,23 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <InternalsVisibleTo Include="Microsoft.Extensions.DependencyModel.Tests" /> 
```
We usually try to avoid this but I understand that you are only moving the code and aren't introducing this.

eerhardt replied:
Yeah, the tests rely heavily on being able to implement internal interfaces for mocking purposes. Exposing those interfaces doesn't make sense, and so IVT was chosen.

joperezr commented on src/libraries/pkg/baseline/packageIndex.json:
```diff
@@ -522,6 +522,32 @@
+        "3.1.1",
+        "3.1.2",
+        "3.1.3"
+      ],
```
Not strictly required but could you add a `"BaselineVersion": "5.0.0",` after the StableVersions as well?

eerhardt replied:
I can do it, but the other Microsoft.Extensions entries don't have it. What does it do?

joperezr replied:
It will make sure that if any other package has a package dependency to this one, the dependency will be on the 5.0.0 package and not a previous version. We usually re-baseline every major release, we already did it for 5.0 but the extensions packages where added after that

joperezr commented on src/libraries/Microsoft.Extensions.DependencyModel/tests/Microsoft.Extensions.DependencyModel.Tests.csproj:
```diff
@@ -0,0 +1,17 @@
+    <PackageReference Include="Microsoft.DotNet.InternalAbstractions" Version="1.0.0" />
+    <PackageReference Include="FluentAssertions" Version="4.19.4" />
+    <PackageReference Include="Moq" Version="$(MoqVersion)" />
+    <PackageReference Include="Microsoft.DotNet.ProjectModel" Version="1.0.0-rc2-002702" />
```
@safern do you think that the versions of these packages should live in Versions.props?

eerhardt replied:
Both `Microsoft.DotNet.InternalAbstractions` and `Microsoft.DotNet.ProjectModel` are obsolete packages (think `project.json` and `.xproj` time frame). The issue is our unit tests are written relying on these old packages, and we never took the time to rewrite the tests. But we don't want to lose the test coverage.

These versions will never be updated, so I didn't think we would want them in a global/central place.

`FluentAssertions` is different - this version could get updated in the future. However since no other test in src\libraries appears to use this package, my hope is we can remove this dependency in the future.

safern replied:
I think it is fine to keep them here. Maybe `FluentAssertions` as you mention might be good to move it, but since you intend to remove it, I don't have an strong opinion.

eerhardt replied:
Do we intend to use `FluentAssertions` more in the repo? If we wanted to use it more, then I'd be OK with moving it. But I was thinking we didn't want to use it (since no one uses it today).

danmosemsft replied:
My impression is that we don't want to - as it's inconsistent with what we've been doing (amongst some other reasons). if we want to, we should have a discussion first and achieve consensus.

joperezr commented:
Left a few comments but this Looks good in general.

eerhardt committed:
Exclude DependencyModel in shims and package checks since it has a duplicated type with System.Collections.

eerhardt committed:
Disable harvesting the stable package.

eerhardt committed:
Respond to PR feedback.

Also, clean up references to the old DependencyModel project.

eerhardt commented:
> Since its just extension methods that folks aren't going to directly reference it's not a terrible type conflict. I'd just suppress the error for now and file a bug. If folks have a scenario for using the two types and notice the clash we can address it.

Done - #34420

>> HashCodeCombiner

>What do you suggest? Removing is breaking, we could obsolete it and/or implement on top of System.HashCode?

My suggestion is to make the breaking change in `5.0` (which I've done in this PR). It was just added in `3.0` (mistakenly since it was a shared source file, and the type is public), and if anyone is using it, they shouldn't be. Also, if they were using it - it would collide with the same type in PlatformAbstractions. So I think we can make this small break, so we don't have to worry about it going forward.

eerhardt committed:
Harvest previous TFM assets from the previously shipped package.

This also means we start building DependencyModel for net461 to ensure full framework support works correctly, and doesn't pick up the old net451 asset.

eerhardt committed:
Fix unit tests to consistently pass on core and netfx.

eerhardt committed:
Add back HashCodeCombiner as obsolete.

joperezr commented on src/libraries/Microsoft.Extensions.DependencyModel/ref/Microsoft.Extensions.DependencyModel.csproj:
```diff
@@ -0,0 +1,11 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;net461</TargetFrameworks>
```
NIT: technically since you are not shipping the ref in the package and the API surface is not changing between netstandard2.0 and net461, it should be fine to only keep the netstandard2.0 configuration in this project if you wanted.

joperezr commented on src/libraries/Microsoft.Extensions.DependencyModel/src/Microsoft.Extensions.DependencyModel.csproj:
```diff
@@ -0,0 +1,29 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFrameworks>netstandard2.0;net461</TargetFrameworks>
```
In order to fix the net472 build, you will also need to include a configuration here for $(NetFrameworkCurrent) and will also make sure you set the property `<ExcludeCurrentFullFrameworkFromPackage>true</ExcludeCurrentFullFrameworkFromPackage>`

eerhardt committed:
Fix netfx build.

eerhardt committed:
Fix DependencyModel pkg build to reference the correct version of Newtonsoft.Json for the harvested package assets.

joperezr committed:
Adding IgnoredTypes for Serialization.Primitives in netcoreapp1.0 and netcoreapp1.1

joperezr reviewed (approved):
Left a couple of small questions, but this looks good to me otherwise. Thanks @eerhardt

eerhardt commented on src/libraries/shims/generated/Directory.Build.props:
```diff
@@ -12,9 +12,10 @@
     <ReferencePath
       Include="$(RefPath)*.dll"
-      Exclude="$(RefPath)$(MSBuildProjectName).dll;$(RefPath)netstandard.dll" />
+      Exclude="$(RefPath)$(MSBuildProjectName).dll;$(RefPath)netstandard.dll;$(RefPath)Microsoft.Extensions.DependencyModel.dll" />
```
I'm not aware of a better fix here. If there is one, I'd be happy to change this.

Honestly, I'm not sure why we are passing `Microsoft.Extensions.*` libraries in here at all.

ericstj replied:
You could, however I'm not sure how much value that has.  We don't intent for any Micrsoft.Extensions assemblies to satisfy .NETFramework nor .NETStandard surface area, so I am fine omitting them.
